<!DOCTYPE html><html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><title>Beginning Python Programming — Part 9</title><style>
      * {
        font-family: Georgia, Cambria, "Times New Roman", Times, serif;
      }
      html, body {
        margin: 0;
        padding: 0;
      }
      h1 {
        font-size: 50px;
        margin-bottom: 17px;
        color: #333;
      }
      h2 {
        font-size: 24px;
        line-height: 1.6;
        margin: 30px 0 0 0;
        margin-bottom: 18px;
        margin-top: 33px;
        color: #333;
      }
      h3 {
        font-size: 30px;
        margin: 10px 0 20px 0;
        color: #333;
      }
      header {
        width: 640px;
        margin: auto;
      }
      section {
        width: 640px;
        margin: auto;
      }
      section p {
        margin-bottom: 27px;
        font-size: 20px;
        line-height: 1.6;
        color: #333;
      }
      section img {
        max-width: 640px;
      }
      footer {
        padding: 0 20px;
        margin: 50px 0;
        text-align: center;
        font-size: 12px;
      }
      .aspectRatioPlaceholder {
        max-width: auto !important;
        max-height: auto !important;
      }
      .aspectRatioPlaceholder-fill {
        padding-bottom: 0 !important;
      }
      header,
      section[data-field=subtitle],
      section[data-field=description] {
        display: none;
      }
      </style></head><body><article class="h-entry">
<header>
<h1 class="p-name">Beginning Python Programming — Part 9</h1>
</header>
<section data-field="subtitle" class="p-summary">
Everything you need to know about error handling in Python
</section>
<section data-field="body" class="e-content">
<section name="6f54" class="section section--body section--first"><div class="section-divider"><hr class="section-divider"></div><div class="section-content"><div class="section-inner sectionLayout--insetColumn"><h3 name="2bfc" id="2bfc" class="graf graf--h3 graf--leading graf--title">Beginning Python Programming — Part 9</h3><h4 name="0187" id="0187" class="graf graf--h4 graf-after--h3 graf--subtitle">Everything you need to know about error handling in Python</h4></div><div class="section-inner sectionLayout--outsetColumn"><figure name="fcc0" id="fcc0" class="graf graf--figure graf--layoutOutsetCenter graf-after--h4"><div class="aspectRatioPlaceholder is-locked" style="max-width: 1032px; max-height: 688px;"><div class="aspectRatioPlaceholder-fill" style="padding-bottom: 66.7%;"></div><img class="graf-image" data-image-id="0*ae55KFpfSHWiWL6c" data-width="5568" data-height="3712" data-unsplash-photo-id="5jqSxgQ6TD4" data-is-featured="true" src="https://cdn-images-1.medium.com/max/1200/0*ae55KFpfSHWiWL6c"></div><figcaption class="imageCaption">Photo by <a href="https://unsplash.com/@chuttersnap?utm_source=medium&amp;utm_medium=referral" data-href="https://unsplash.com/@chuttersnap?utm_source=medium&amp;utm_medium=referral" class="markup--anchor markup--figure-anchor" rel="photo-creator noopener noopener noopener noopener noopener noopener noopener noopener" target="_blank">chuttersnap</a> on <a href="https://unsplash.com?utm_source=medium&amp;utm_medium=referral" data-href="https://unsplash.com?utm_source=medium&amp;utm_medium=referral" class="markup--anchor markup--figure-anchor" rel="photo-source noopener noopener noopener noopener noopener noopener noopener noopener" target="_blank">Unsplash</a></figcaption></figure></div><div class="section-inner sectionLayout--insetColumn"><p name="148e" id="148e" class="graf graf--p graf-after--figure">In the last post, we covered formatting, list comprehensions, and lambdas.</p><div name="8976" id="8976" class="graf graf--mixtapeEmbed graf-after--p"><a href="https://medium.com/@broebling/beginning-python-programming-part-8-966f1189c1fe" data-href="https://medium.com/@broebling/beginning-python-programming-part-8-966f1189c1fe" class="markup--anchor markup--mixtapeEmbed-anchor" title="https://medium.com/@broebling/beginning-python-programming-part-8-966f1189c1fe"><strong class="markup--strong markup--mixtapeEmbed-strong">Beginning Python Programming — Part 8</strong><br><em class="markup--em markup--mixtapeEmbed-em">In the last post, we dug deep into classes and discovered how they were created and destroyed. We also looked at how we…</em>medium.com</a><a href="https://medium.com/@broebling/beginning-python-programming-part-8-966f1189c1fe" class="js-mixtapeImage mixtapeImage u-ignoreBlock" data-media-id="ba5a15d209b2efb5460d53d21efa9953" data-thumbnail-img-id="0*l5XIwT5itNQSmAxn" style="background-image: url(https://cdn-images-1.medium.com/fit/c/160/160/0*l5XIwT5itNQSmAxn);"></a></div><p name="6b36" id="6b36" class="graf graf--p graf-after--mixtapeEmbed">With all that flexibility, we also made our code complex. When you have a sophisticated code base, you are sure to run into errors, and this is where error handling comes in to play.</p><p name="f4ee" id="f4ee" class="graf graf--p graf-after--p">First and foremost, error handling is <strong class="markup--strong markup--p-strong"><em class="markup--em markup--p-em">not</em></strong> testing! Nor is the print statement for that matter. So what exactly is error handling?</p><h3 name="68bb" id="68bb" class="graf graf--h3 graf-after--p">Error Handling</h3><p name="5b7c" id="5b7c" class="graf graf--p graf-after--h3">Error handling, in short, is adding enough logic so that your program can handle cases where it would otherwise stop running.</p><p name="fdba" id="fdba" class="graf graf--p graf-after--p">Errors come from functions that can raise errors. When an error has risen, it is returned from a function, even if the function is supposed to return a string value. If uncaught, it will cause your program to crash. So let’s get started so we can handle these effectively.</p><p name="c7a0" id="c7a0" class="graf graf--p graf-after--p">Let’s look at a few examples where errors are easy to spot.</p><figure name="41be" id="41be" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/d54e67c28ff93caf025eb00040dc4522.js"></script></figure><p name="2b86" id="2b86" class="graf graf--p graf-after--figure">First, we are trying to access the <code class="markup--code markup--p-code">address</code> key from <code class="markup--code markup--p-code">my_dict</code>. As expected, that key does not exist, and we receive a <code class="markup--code markup--p-code">KeyError</code> in our console.</p><p name="5b4b" id="5b4b" class="graf graf--p graf-after--p">Second, we try to access the 10th element in <code class="markup--code markup--p-code">my_list</code> which also doesn’t exist. We receive an <code class="markup--code markup--p-code">IndexError</code> for this scenario.</p><p name="8a48" id="8a48" class="graf graf--p graf-after--p">We can clear this up using <code class="markup--code markup--p-code">try...except</code>.</p><figure name="31b7" id="31b7" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/1ffc42ea4e1773d5d29697a3834bdde3.js"></script></figure><p name="6c56" id="6c56" class="graf graf--p graf-after--figure">I’ve slightly changed the order, but it effectively does the same thing as before, only I’ve surrounded the failing code from earlier in a try/except block.</p><p name="62f6" id="62f6" class="graf graf--p graf-after--p">What happens is we prepare the Python interpreter for potentially failing code by writing <code class="markup--code markup--p-code">try:</code>. Inside of the <code class="markup--code markup--p-code">try</code> body Python will check for errors as it runs each line. If an error occurs, it steps out of the <code class="markup--code markup--p-code">try</code> body and performs the code written inside of <code class="markup--code markup--p-code">except:</code>.</p><p name="468e" id="468e" class="graf graf--p graf-after--p">Our program no longer crashes, but it does print “whoops!” and “whoops again!” to the console. This is good because previously our program would exit as soon as it hit the first failing <code class="markup--code markup--p-code">print</code> statement, now our program continues. If we had more code written below the last try/except statement, that code would run as well.</p><p name="290b" id="290b" class="graf graf--p graf-after--p">I’m not going to stop there, because while that code works, it’s not <em class="markup--em markup--p-em">good</em> code.</p><p name="b3b3" id="b3b3" class="graf graf--p graf-after--p">Let’s rewrite this a correct way.</p><figure name="5912" id="5912" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/ad132213e522552c25db5b041b0fe607.js"></script></figure><p name="66ee" id="66ee" class="graf graf--p graf-after--figure">Here we add the type of error that we are expecting to receive from the failed code next to <code class="markup--code markup--p-code">except</code>. The cool thing about most errors across all languages is they tend to tell you which error occurred followed by a reason for the occurrence.</p><p name="98ac" id="98ac" class="graf graf--p graf-after--p">We can still improve on this by capturing the error to display to users or log to a console.</p><figure name="9c85" id="9c85" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/6a899b4f593756f101ae6c1128241787.js"></script></figure><p name="7c51" id="7c51" class="graf graf--p graf-after--figure">Now we are getting back useful information about the error that occurred. With proper logger formatting (later article), we can include the line number that failed, or even provide more information if the number 10 was stored in a variable.</p><p name="7859" id="7859" class="graf graf--p graf-after--p">In scenarios where <code class="markup--code markup--p-code">my_dict</code> is required to return a value from <code class="markup--code markup--p-code">my_list</code>, we can wrap them up in the same <code class="markup--code markup--p-code">try</code> block, if one fails, then we stop processing the rest of the <code class="markup--code markup--p-code">try</code> body, perform the <code class="markup--code markup--p-code">except</code> body and continue.</p><figure name="7b2f" id="7b2f" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/e28c4a128584bba757027d0068c2e24f.js"></script></figure><p name="8365" id="8365" class="graf graf--p graf-after--figure">Since this failed to retrieve the dictionary value, it didn’t even try to retrieve the list value. It dropped down to our final <code class="markup--code markup--p-code">print</code> statement and the program finished.</p><blockquote name="c85f" id="c85f" class="graf graf--blockquote graf-after--p">This can be helpful in the case of network-related calls where the person may not have Internet access. Since we usually have more code that is dependent on this network access we can skip over it and save some processing time. This is just an example, a more realistic solution to network failures is to “early exit” from whatever function is performing the network call, that is, just return from the function as soon as the call fails.</blockquote><p name="10ef" id="10ef" class="graf graf--p graf-after--blockquote">Note that when you print an error, you must cast it as a string. Since print automatically calls <code class="markup--code markup--p-code">__str__(self)</code> within any class for you, we only need to do this when appending the error to a string. We can also use f-strings in place of this, which will also do this for us.</p><p name="c98f" id="c98f" class="graf graf--p graf-after--p">There is one more thing I should add to ensure this is fail-proof. That is a global exception clause.</p><p name="9743" id="9743" class="graf graf--p graf-after--p">All exceptions derive from the <code class="markup--code markup--p-code">BaseException</code> and <code class="markup--code markup--p-code">Exception</code> classes. The difference between the two is easy enough.</p><ul class="postList"><li name="c9a0" id="c9a0" class="graf graf--li graf-after--p"><code class="markup--code markup--li-code">BaseException</code> is where all exceptions are subclassed from, including <code class="markup--code markup--li-code">Exception</code></li><li name="da78" id="da78" class="graf graf--li graf-after--li"><code class="markup--code markup--li-code">Exception</code> is where all non-exit exceptions subclass. According to the <a href="https://docs.python.org/3.7/library/exceptions.html#Exception" data-href="https://docs.python.org/3.7/library/exceptions.html#Exception" class="markup--anchor markup--li-anchor" rel="noopener" target="_blank">python documentation</a>: “All user-defined exceptions should also be derived from this class.”</li></ul><figure name="44a0" id="44a0" class="graf graf--figure graf--iframe graf-after--li"><script src="https://gist.github.com/mildocjr/f207066950523939adf9a301eda627cd.js"></script></figure><p name="fd83" id="fd83" class="graf graf--p graf-after--figure">I just added another exception of <code class="markup--code markup--p-code">BaseException</code>. This way if anything we weren’t expecting comes through, we will surely handle it and receive back a message that gives us a good hint on what went wrong.</p><p name="8296" id="8296" class="graf graf--p graf-after--p">I’ve also made another change to the <code class="markup--code markup--p-code">IndexError</code> exception, if the number is out of range, we print the last item in the list. While this isn’t fool-proof, we are more likely to want the last item in the list than the first, which is always at index 0; therefore, this is a good enough solution. Consider this a foray into <em class="markup--em markup--p-em">heuristic programming, </em>or more simply stated best guess logic<em class="markup--em markup--p-em">.</em></p><p name="0374" id="0374" class="graf graf--p graf-after--p">This is the real purpose of exceptions, and we don’t just want to display them, we want to act on them, we need to do stuff with the error and make it meaningful to the user. The user doesn’t care how many elements are in the list; they know that if they want the last one, they are going to ask for a high value from that list and hope it works.</p><p name="1ead" id="1ead" class="graf graf--p graf-after--p">Likewise, we could substitute the <code class="markup--code markup--p-code">KeyError</code> exception with something like this:</p><figure name="3b99" id="3b99" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/eb10e7669ee24e3a75ff8974bdd23e63.js"></script></figure><p name="4bc7" id="4bc7" class="graf graf--p graf-after--figure">See how helpful that is? Make your errors meaningful.</p><h4 name="eb1a" id="eb1a" class="graf graf--h4 graf-after--p">Else</h4><p name="e6da" id="e6da" class="graf graf--p graf-after--h4">Sometimes we might want to do something only if everything goes well. We can add <code class="markup--code markup--p-code">else</code> to our <code class="markup--code markup--p-code">try/except</code> and this will only execute if we didn’t have any issues.</p><figure name="2a8c" id="2a8c" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/8993db6238ef60d6461855a89999733c.js"></script></figure><p name="48cd" id="48cd" class="graf graf--p graf-after--figure">Here we are trying to open a file named <code class="markup--code markup--p-code">my_file.txt</code> as read-only, denoted by <code class="markup--code markup--p-code">&#39;r&#39;</code>. If successful, we read the file’s contents and store it in <code class="markup--code markup--p-code">contents</code> before printing the contents of the file to the console.</p><p name="8e1e" id="8e1e" class="graf graf--p graf-after--p">If the file does not exist in the same location as our running program, we will hit a <code class="markup--code markup--p-code">FileNotFoundError</code>. If that happens, we tell the user about what happened and move on.</p><p name="8559" id="8559" class="graf graf--p graf-after--p">If this doesn’t happen, we have an open file on our hands, and we need to ensure we close the file to remove any locks so other programs can use the file later. This is where <code class="markup--code markup--p-code">else</code> comes in. If it is successful, the else clause executes, and we close the file.</p><p name="6cbe" id="6cbe" class="graf graf--p graf-after--p">This is working code, so if you want to make a new file called “my_file.txt” in your working directory this will work, and the file will close when done.</p><p name="67af" id="67af" class="graf graf--p graf-after--p"><em class="markup--em markup--p-em">In Python, files are automatically closed for you by garbage collection or when your program exits; however, it is always better to explicitly close any files that you open. There is a better way to handle file IO, but we will cover that in a future article.</em></p><h4 name="f374" id="f374" class="graf graf--h4 graf-after--p">Finally</h4><p name="260a" id="260a" class="graf graf--p graf-after--h4">No this isn’t the end of the article, but there is a sibling to <code class="markup--code markup--p-code">else</code> called <code class="markup--code markup--p-code">finally</code>. The difference between the two is pretty easy to understand.</p><ul class="postList"><li name="afc1" id="afc1" class="graf graf--li graf-after--p"><code class="markup--code markup--li-code">else</code> is <strong class="markup--strong markup--li-strong">only</strong> called when the try statement is successful.</li><li name="c189" id="c189" class="graf graf--li graf-after--li"><code class="markup--code markup--li-code">finally</code> is <strong class="markup--strong markup--li-strong">always</strong> called no matter what.</li></ul><p name="9e64" id="9e64" class="graf graf--p graf-after--li">They can be used separately or together, depending on your need.</p><figure name="272a" id="272a" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/90efcccda1a71fe60a8aa9c1edc15647.js"></script></figure><p name="76d7" id="76d7" class="graf graf--p graf-after--figure">If successful, we will see the following output:</p><pre name="6dda" id="6dda" class="graf graf--pre graf-after--p">file opened!<br>Hello, World!<br>file closed<br>done</pre><p name="249f" id="249f" class="graf graf--p graf-after--pre">If the file was not found:</p><pre name="442c" id="442c" class="graf graf--pre graf-after--p">File not found!<br>done</pre><p name="18da" id="18da" class="graf graf--p graf-after--pre">An excellent way to start writing your <code class="markup--code markup--p-code">try/except</code> blocks is to begin by separating everything away from each other. Figure out what kinds of errors might come from error raising functions.</p></div><div class="section-inner sectionLayout--outsetColumn"><figure name="55e4" id="55e4" class="graf graf--figure graf--layoutOutsetCenter graf-after--p"><div class="aspectRatioPlaceholder is-locked" style="max-width: 1032px; max-height: 688px;"><div class="aspectRatioPlaceholder-fill" style="padding-bottom: 66.7%;"></div><img class="graf-image" data-image-id="0*B-qzw1RcHERcC_Jf" data-width="5308" data-height="3538" data-unsplash-photo-id="Bjw6dbIYtNE" src="https://cdn-images-1.medium.com/max/1200/0*B-qzw1RcHERcC_Jf"></div><figcaption class="imageCaption">Photo by <a href="https://unsplash.com/@peterhershey?utm_source=medium&amp;utm_medium=referral" data-href="https://unsplash.com/@peterhershey?utm_source=medium&amp;utm_medium=referral" class="markup--anchor markup--figure-anchor" rel="photo-creator noopener noopener" target="_blank">Peter Hershey</a> on <a href="https://unsplash.com?utm_source=medium&amp;utm_medium=referral" data-href="https://unsplash.com?utm_source=medium&amp;utm_medium=referral" class="markup--anchor markup--figure-anchor" rel="photo-source noopener noopener" target="_blank">Unsplash</a></figcaption></figure></div><div class="section-inner sectionLayout--insetColumn"><h3 name="73d1" id="73d1" class="graf graf--h3 graf-after--figure">Creating Custom Errors</h3><p name="619f" id="619f" class="graf graf--p graf-after--h3">Sometimes our program is so specialized that we need errors of our own that we can return from functions that provide the specificity to tell our user what went wrong. We have two choices, either adopt an error that Python already has, or create one of our own.</p><p name="e152" id="e152" class="graf graf--p graf-after--p">Let’s look at adopting from Python first.</p><p name="8a19" id="8a19" class="graf graf--p graf-after--p">For this example, we are going to accept a string parameter, but we expect it only to contain spaces.</p><figure name="1cf2" id="1cf2" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/5e4dc6af5c89349f89382819fe35ffb8.js"></script></figure><p name="cd2b" id="cd2b" class="graf graf--p graf-after--figure">In the two examples, the first <code class="markup--code markup--p-code">data</code> value is correct, and we see a value of 10 displayed in our console.</p><p name="2a44" id="2a44" class="graf graf--p graf-after--p">In the second example, we have tabs (<code class="markup--code markup--p-code">\t</code>) in place. The first thing our function does is check if data contains tabs. If it does it raises a <code class="markup--code markup--p-code">TabError</code>. The error itself is disconnected from the <code class="markup--code markup--p-code">data</code> string; it doesn’t care about the contents. It’s just an error that you can use to provide a meaningful stop in your program if things don’t go as planned. If you try printing the string value of <code class="markup--code markup--p-code">TabError</code> you will get back <code class="markup--code markup--p-code">None</code>.</p><p name="6629" id="6629" class="graf graf--p graf-after--p">If you wanted to include an error message more tailored to the situation at hand, you could change the function to read as such:</p><figure name="6ee2" id="6ee2" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/ed2038754a559aaa7ae880b71e579908.js"></script></figure><p name="196d" id="196d" class="graf graf--p graf-after--figure">This would allow you to use <code class="markup--code markup--p-code">TabError as error</code> like we did above and display that as an error message instead. This is preferred, this way, you only need to worry about returning your error message instead of creating one every time you call this function.</p><p name="3830" id="3830" class="graf graf--p graf-after--p graf--trailing">Of course, if you need logic in your error message, you can adjust one or both messages to meet your needs.</p></div></div></section><section name="008f" class="section section--body section--last"><div class="section-divider"><hr class="section-divider"></div><div class="section-content"><div class="section-inner sectionLayout--insetColumn"><p name="bcb6" id="bcb6" class="graf graf--p graf--leading">If you need something even more special than this, you can always create your own error.</p><figure name="5d34" id="5d34" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/54d15cd34e9681fe2ea688afe6548796.js"></script></figure><p name="35b0" id="35b0" class="graf graf--p graf-after--figure">Here we create an exception named <code class="markup--code markup--p-code">BadData</code> it subclasses from <code class="markup--code markup--p-code">Exception</code> which is what we should be using for all of our errors. I added in a <em class="markup--em markup--p-em">docstring</em> which helps us understand what it should be used for when we see the automatic documentation for <code class="markup--code markup--p-code">BadData</code>. Then we <code class="markup--code markup--p-code">pass</code> because we don’t need to do anything more.</p><p name="87f8" id="87f8" class="graf graf--p graf-after--p">We run through our program, and if we have a file containing levels, then we continue processing the contents. If not, then we raise our exception and catch it below. Because our exception is a class, we can use it anywhere it is needed.</p><p name="667b" id="667b" class="graf graf--p graf-after--p">It’s straightforward and provides a lot of functionality for you to handle errors that the system may not understand. Always create custom exceptions if you cannot find one suitable for your needs and there is no better way to refactor your code to handle that error without using <code class="markup--code markup--p-code">try/except</code>.</p><h4 name="6664" id="6664" class="graf graf--h4 graf-after--p">Re-Raise</h4><p name="20f0" id="20f0" class="graf graf--p graf-after--h4">Sometimes you might call an exception deep in a stack of functions and need it to bubble back up to the top. In scenarios like this, you can use <code class="markup--code markup--p-code">raise</code> to continue raising the same exception back up through the stack until you are ready to handle it.</p><figure name="4531" id="4531" class="graf graf--figure graf--iframe graf-after--p"><script src="https://gist.github.com/mildocjr/735f3ce9ef47e50ce677a36b2cd8e809.js"></script></figure><p name="d144" id="d144" class="graf graf--p graf-after--figure">Here I’m just using a <code class="markup--code markup--p-code">NotImplementedError</code> because well, we haven’t completed the function yet. <code class="markup--code markup--p-code">some_other_func</code> calls <code class="markup--code markup--p-code">make_error</code> using a <code class="markup--code markup--p-code">try/except</code> block. If we receive a <code class="markup--code markup--p-code">NotImplementedError</code> we raise it again so we can handle it in our main body where we print “Not implemented.” While this is a simple example, it gives you a good idea of how it works.</p><h3 name="f998" id="f998" class="graf graf--h3 graf-after--p">Summary</h3><p name="4b56" id="4b56" class="graf graf--p graf-after--h3">Today we learned how errors occur and how we can handle them if they do. We learned several techniques for keeping your program running under less than ideal conditions as well as differences between proper error handling and lousy error handling. If you’ve downloaded PyCharm, you can type <code class="markup--code markup--p-code">raise</code>, then <code class="markup--code markup--p-code">ctrl</code> + <code class="markup--code markup--p-code">space</code> to see some of the exceptions that you can use. If you have been brave enough to get involved with modules from other places, you can even see the different exceptions they have created for you concerning their projects.</p><h4 name="f786" id="f786" class="graf graf--h4 graf-after--p">Suggested Reading</h4><p name="da53" id="da53" class="graf graf--p graf-after--h4">Chapter 8 of the Python Tutorial</p><div name="a54a" id="a54a" class="graf graf--mixtapeEmbed graf-after--p"><a href="https://docs.python.org/3/tutorial/errors.html" data-href="https://docs.python.org/3/tutorial/errors.html" class="markup--anchor markup--mixtapeEmbed-anchor" title="https://docs.python.org/3/tutorial/errors.html"><strong class="markup--strong markup--mixtapeEmbed-strong">8. Errors and Exceptions — Python 3.7.3 documentation</strong><br><em class="markup--em markup--mixtapeEmbed-em">Even if a statement or expression is syntactically correct, it may cause an error when an attempt is made to execute…</em>docs.python.org</a><a href="https://docs.python.org/3/tutorial/errors.html" class="js-mixtapeImage mixtapeImage mixtapeImage--empty u-ignoreBlock" data-media-id="418c6ed6fb2ee834de124b33ab45f574"></a></div><h3 name="eb37" id="eb37" class="graf graf--h3 graf-after--mixtapeEmbed">What’s Next</h3><p name="0b81" id="0b81" class="graf graf--p graf-after--h3">Code Structure and Readability is what’s next. It’s about time that we learn how to structure our code under PEP 8 standards properly. We might also cover docstring conventions defined in PEP 257.</p><p name="9469" id="9469" class="graf graf--p graf-after--p">If you are going to write good Python programs, you need to understand those PEP standards and abide by them at a minimum.</p><p name="a1c3" id="a1c3" class="graf graf--p graf-after--p graf--trailing">Until then, see if you can handle pesky errors that may have popped up in some of your projects. Make it so that you can throw whatever data you want at your application and it intelligently responds to your misgivings.</p></div></div></section>
</section>
<footer><p>By <a href="https://medium.com/@broebling" class="p-author h-card">Bob Roebling</a> on <a href="https://medium.com/p/45cad890e6b"><time class="dt-published" datetime="2019-06-04T01:56:35.812Z">June 4, 2019</time></a>.</p><p><a href="https://medium.com/@broebling/beginning-python-programming-part-8-45cad890e6b" class="p-canonical">Canonical link</a></p><p>Exported from <a href="https://medium.com">Medium</a> on October 1, 2019.</p></footer></article></body></html>